service/history/failover/coordinator.go

```diff
@@ -0,0 +1,297 @@
+		notificationChan: make(chan *notificationRequest, notificationChanBufferSize),
+		receiveChan:      make(chan *receiveRequest, receiveChanBufferSize),
+		shutdownChan:     make(chan struct{}),
+		retryPolicy:      retryPolicy,
+		metadataMgr:      metadataMgr,
+		historyClient:    historyClient,
+		timeSource:       timeSource,
+		config:           config,
+		metrics:          metrics,
+		logger:           logger,
+	}
+}
+
+func (c *coordinatorImpl) Start() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	go c.receiveFailoverMarkersLoop()
+	go c.notifyFailoverMarkerLoop()
+
+	c.logger.Info("Failover coordinator started.")
+}
+
+func (c *coordinatorImpl) Stop() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(c.shutdownChan)
+	c.logger.Info("Failover coordinator stopped.")
+}
+
+func (c *coordinatorImpl) NotifyFailoverMarkers(
+	shardID int,
+	markers []*replicator.FailoverMarkerAttributes,
+) <-chan error {
+
+	respCh := make(chan error, 1)
+	c.notificationChan <- &notificationRequest{
+		shardID: shardID,
+		markers: markers,
+		respCh:  respCh,
+	}
+
+	return respCh
+}
+
+func (c *coordinatorImpl) ReceiveFailoverMarkers(
+	shardIDs []int,
+	marker *replicator.FailoverMarkerAttributes,
+) {
+
+	c.receiveChan <- &receiveRequest{
+		shardIDs: shardIDs,
+		marker:   marker,
+	}
+}
+
+func (c *coordinatorImpl) receiveFailoverMarkersLoop() {
+
+	ticker := time.NewTicker(cleanupMarkerInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.shutdownChan:
+			return
+		case <-ticker.C:
+			c.cleanupInvalidMarkers()
+		case request := <-c.receiveChan:
+			c.handleFailoverMarkers(request)
+		}
+	}
+}
+
+func (c *coordinatorImpl) notifyFailoverMarkerLoop() {
+
+	timer := time.NewTimer(backoff.JitDuration(
+		c.config.FailoverMarkerHeartbeatInterval(),
+		c.config.FailoverMarkerHeartbeatTimerJitterCoefficient(),
+	))
+	defer timer.Stop()
+	requestByMarker := make(map[*replicator.FailoverMarkerAttributes]*receiveRequest)
+
+	for {
+		select {
+		case <-c.shutdownChan:
+			return
+		case notificationReq := <-c.notificationChan:
+			// if there is a shard movement happen, it is fine to have duplicate shard ID in the request
+			// The receiver side will de-dup the shard IDs. See: handleFailoverMarkers
+			for _, marker := range notificationReq.markers {
+				if _, ok := requestByMarker[marker]; !ok {
+					requestByMarker[marker] = &receiveRequest{
+						shardIDs: []int{},
+						marker:   marker,
+						respCh:   notificationReq.respCh,
+					}
+				}
+				req := requestByMarker[marker]
+				req.shardIDs = append(req.shardIDs, notificationReq.shardID)
+			}
+		case <-timer.C:
+			if len(requestByMarker) > 0 {
+				var err error
+				//TODO: Remote calls to send failover markers
+				for marker, request := range requestByMarker {
+					request.respCh <- err
```

Comment:
What the expected behavior of caller when it receives an error through the channel? Retry? But I thought even if the notify succeeds, it still needs to periodically call notify in case the `CleanPendingActiveState` fails?

Reply:
Yes. The error won't change the behavior. The coordinator will continue to notify the failover markers. The channel is to let the goroutine wait for the response and schedule next notify call.